src/main/scala/io/tabmo/aerospike/client/operations/ShapeQuery.scala

```diff
@@ -0,0 +1,111 @@
+package io.tabmo.aerospike.client.operations
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+
+import com.aerospike.client.{AerospikeException, Key, Record, Value}
+import com.aerospike.client.listener.RecordListener
+import io.tabmo.aerospike.TSafe.VRestriction
+import io.tabmo.aerospike.client.ReactiveAerospikeClient
+import io.tabmo.aerospike.data.{AerospikeKeyConverter, AerospikeRecord}
+import io.tabmo.aerospike.utils.{BinShape, Shape}
+
+import shapeless.HList
+import shapeless.UnaryTCConstraint._
+
+trait ShapeQuery {
+
+  protected def prepareQueryEquals[L <: HList : *->*[BinShape]#λ, LR <: HList, T, V: VRestriction](shape: Shape[L, LR, T], filterBin: BinShape[V])(implicit keyConverter: AerospikeKeyConverter[String]): PreparedEqualQuery[T, V] = {
+    new PreparedEqualQuery[T, V] {
+
+      override def apply(value: V)(implicit ctx: ExecutionContext, client: ReactiveAerospikeClient): Future[Seq[T]] = {
+        client.queryEqual(shape.ns, shape.set, shape.binNames(), filterBin.name, value).map {
+          _.values.map { r =>
+            shape.unshape(r)
+          }.toSeq
+        }
+      }
+    }
+  }
+
+  protected def prepareQueryRange[L <: HList : *->*[BinShape]#λ, LR <: HList, T](shape: Shape[L, LR, T], filterBin: BinShape[Long])(implicit keyConverter: AerospikeKeyConverter[String]): PreparedRangeQuery[T] = {
+    new PreparedRangeQuery[T] {
+
+      override def apply(min: Long, max: Long)(implicit ctx: ExecutionContext, client: ReactiveAerospikeClient): Future[Seq[T]] = {
+        client.queryRange(shape.ns, shape.set, shape.binNames(), filterBin.name, min, max).map {
+          _.values.map { r =>
+            shape.unshape(r)
+          }.toSeq
+        }
+      }
+    }
+  }
+
+  protected def prepareQueryIndex[L <: HList : *->*[BinShape]#λ, LR <: HList, T](shape: Shape[L, LR, T]) = {
+    val binsName = shape.binNames()
+    new PreparedIndexQuery[T] {
+      override def apply[V](k: V)(implicit ctx: ExecutionContext, client: ReactiveAerospikeClient): Future[Option[T]] = {
+
+        val key = k match {
```

Comment:
sauf erreur de ma part, ça doit lever un "not exhaustive matching" ce bloc non ? le type V n'est pas limité